File: recursion/cxx/combinations.cxx
#include <vector>

void comb_helper(std::vector<int> nums, int k, int start, std::vector<int>& buf, std::vector<std::vector<int> >& result)
{
	if(buf.size() == k)
	{
		result.push_back(buf);
		return;
	}
	for(int i = start; i < nums.size() - (k - buf.size()) + 1; i++)
	{
		buf.push_back(nums[i]);
		comb_helper(nums, k, i + 1, buf, result);
		buf.pop_back();
	}
}

std::vector<std::vector<int> > combinations(std::vector<int> nums, int k)
{
	std::vector<std::vector<int> > result;
	std::vector<int> buf;
	comb_helper(nums, k, 0, buf, result);
	return result;
}

void comb_with_dup_helper(std::vector<int> nums, int k, int start, std::vector<int>& buf, std::vector<std::vector<int> >& result)
{
	if(buf.size() == k)
	{
		result.push_back(buf);
		return;
	}
	for(int i = start; i < nums.size() - (k - buf.size()) + 1; i++)
		if(i == 0 or nums[i] != nums[i - 1])
		{
			buf.push_back(nums[i]);
			comb_with_dup_helper(nums, k, i + 1, buf, result);
			buf.pop_back();
		}
}

std::vector<std::vector<int> > combinations_with_dup(std::vector<int> nums, int k)
{
	std::vector<std::vector<int> > result;
	std::vector<int> buf;
	comb_with_dup_helper(nums, k, 0, buf, result);
	return result;
}

File: recursion/py3/combinations.py
# Combination C(i, n)
"""
Same algorithm with subsets. Only different on stop conditions
"""
def combinations(nums, k):
    def helper(k, start, buf, result):
        if len(buf) == k:
            result.append(list(buf))
            return result
        for i in range(start, len(nums) - (k - len(buf)) + 1):
            buf.append(nums[i])
            helper(k, i + 1, buf, result)
            buf.pop()
        return result
    return helper(k, 0, [], [])

def combinationsWithDup(nums, k):
    def helper(k, start, buf, result):
        if len(buf) == k:
            result.append(list(buf))
            return result
        for i in range(start, len(nums) - (k - len(buf)) + 1):
            if i == start or nums[i] != nums[i - 1]:
                buf.append(nums[i])
                helper(k, i + 1, buf, result)
                buf.pop()
        return result
    return helper(k, 0, [], [])
